// urequire/build/code/config/MasterDefaultsConfig.js
// Generated by CoffeeScript 1.6.3
var MasterDefaultsConfig;

module.exports = MasterDefaultsConfig = {
  bundle: {
    path: void 0,
    filez: void 0,
    copy: [],
    name: void 0,
    main: void 0,
    dependencies: {
      exports: {
        bundle: {},
        root: {},
        replace: void 0
      },
      node: ['fs', 'events', 'util', 'http', 'path', 'child_process', 'events', 'crypto', 'string_decoder', 'timers', 'tls', 'domain', 'buffer', 'net', 'dgram', 'dns', 'stream', 'https', 'querystring', 'punycode', 'readline', 'url', 'repl', 'vm', 'assert', 'tty', 'zlib', 'os', 'cluster'],
      locals: {},
      depsVars: {},
      _knownDepsVars: {
        chai: 'chai',
        mocha: 'mocha',
        lodash: "_",
        underscore: "_",
        jquery: ["$", "jQuery"],
        backbone: "Backbone",
        knockout: ["ko", 'Knockout']
      }
    },
    resources: require('./ResourceConverters').defaultResourceConverters,
    commonCode: void 0,
    webRootMap: '.'
  },
  build: {
    dstPath: void 0,
    forceOverwriteSources: false,
    template: 'UMD',
    runtimeInfo: true,
    bare: false,
    useStrict: void 0,
    globalWindow: true,
    injectExportsModule: true,
    noRootExports: false,
    scanAllow: false,
    allNodeRequires: false,
    dummyParams: false,
    watch: false,
    verbose: false,
    debugLevel: 0,
    "continue": false,
    optimize: false,
    _optimizers: ['uglify2', 'uglify'],
    out: void 0,
    done: function(doneVal) {
      return console.log("done() is missing and I got a " + doneVal + " on the default done()");
    },
    clean: void 0,
    rjs: void 0
  }
};
