// urequire/build/code/codeUtils/toCode.js
// Generated by CoffeeScript 1.6.3
var UError, dfb, escodegen, l, toAST, toCode, _, _B;

_ = (_B = require('uberscore'))._;

l = new _B.Logger('uRequire/codeUtils/toCode');

escodegen = require('escodegen');

dfb = new _B.DeepDefaultsBlender;

UError = require('../utils/UError');

toAST = require('./toAST');

module.exports = toCode = function(astCode, options) {
  var ast, err;
  if (options == null) {
    options = toCode.options;
  }
  if (_.isEmpty(astCode)) {
    return '';
  }
  if (options !== toCode.options) {
    dfb.blend(options, toCode.options);
  }
  ast = toAST(astCode, options.type);
  try {
    return escodegen.generate(ast, options);
  } catch (_error) {
    err = _error;
    throw new UError("toCode: Error generating code from AST with escodegen: " + err + ".", {
      nested: err
    });
  }
};

toCode.options = {
  format: {
    indent: {
      style: '  ',
      base: 0
    },
    json: false,
    renumber: false,
    hexadecimal: false,
    quotes: 'double',
    escapeless: true,
    compact: false,
    parentheses: true,
    semicolons: true
  }
};
