// urequire/build/code/templates/ModuleGeneratorTemplates.js
// Generated by CoffeeScript 1.6.3
var ModuleGeneratorTemplates, Template, isTrueOrFileInSpecs, l, _, _B,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

_ = (_B = require('uberscore'))._;

l = new _B.Logger('uRequire/ModuleGeneratorTemplates');

isTrueOrFileInSpecs = require('../config/isTrueOrFileInSpecs');

Template = require('./Template');

ModuleGeneratorTemplates = (function(_super) {
  __extends(ModuleGeneratorTemplates, _super);

  ModuleGeneratorTemplates.prototype.scope = 'module';

  function ModuleGeneratorTemplates(module) {
    this.module = module;
    ModuleGeneratorTemplates.__super__.constructor.apply(this, arguments);
  }

  Object.defineProperties(ModuleGeneratorTemplates.prototype, {
    bundle: {
      get: function() {
        return this.module.bundle;
      }
    },
    build: {
      get: function() {
        return this.bundle.build;
      }
    },
    isCombined: {
      get: function() {
        return this.build.template.name === 'combined';
      }
    },
    moduleNamePrint: {
      get: function() {
        if (this.module.name) {
          return "'" + this.module.name + "', ";
        } else {
          return "";
        }
      }
    },
    isInjectExportsModule: {
      get: function() {
        var _ref, _ref1, _ref2, _ref3, _ref4, _ref5;
        return (this.module.kind === 'nodejs') || (this.module.isInjectExportsModule && !(((_ref = this.module.defineArrayDeps) != null ? (_ref1 = _ref[0]) != null ? typeof _ref1.isEqual === "function" ? _ref1.isEqual('exports') : void 0 : void 0 : void 0) && (((_ref2 = this.module.parameters) != null ? _ref2[0] : void 0) === 'exports') && ((_ref3 = this.module.defineArrayDeps) != null ? (_ref4 = _ref3[1]) != null ? typeof _ref4.isEqual === "function" ? _ref4.isEqual('module') : void 0 : void 0 : void 0) && (((_ref5 = this.module.parameters) != null ? _ref5[1] : void 0) === 'module')));
      }
    },
    injectExportsModuleParamsPrint: {
      get: function() {
        if (this.isInjectExportsModule) {
          return ', exports, module';
        } else {
          return '';
        }
      }
    },
    parametersPrint: {
      get: function() {
        var par;
        return "require" + this.injectExportsModuleParamsPrint + (((function() {
          var _i, _len, _ref, _results;
          _ref = this.module.parameters;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            par = _ref[_i];
            _results.push(", " + par);
          }
          return _results;
        }).call(this)).join(''));
      }
    },
    defineArrayDepsPrint: {
      get: function() {
        var dep;
        return (_.isEmpty(this.module.defineArrayDeps) ? "" : this.isInjectExportsModule ? "['require', 'exports', 'module'" : "['require'") + ((function() {
          var _i, _len, _ref, _results;
          _ref = this.module.defineArrayDeps;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            dep = _ref[_i];
            _results.push(", " + (dep.name({
              quote: true
            })));
          }
          return _results;
        }).call(this)).join('') + (_.isEmpty(this.module.defineArrayDeps) ? '' : '], ');
      }
    },
    runtimeInfo: {
      get: function() {
        if (this.module.isRuntimeInfo) {
          return Template.prototype.runtimeInfo;
        } else {
          return '';
        }
      }
    },
    useStrictModule: {
      get: function() {
        if (this.module.isUseStrict && !(this.isCombined && _B.isTrue(this.build.useString))) {
          return "'use strict';";
        } else {
          return '';
        }
      }
    },
    isRootExports: {
      get: function() {
        return !(this.module.isNoRootExports || _.isEmpty(this.module.flags.rootExports));
      }
    }
  });

  /* private*/


  ModuleGeneratorTemplates.prototype._rootExportsNoConflict = function(rootName, returnModule) {
    var exp, exportedVar, i;
    if (rootName == null) {
      rootName = 'root';
    }
    if (returnModule == null) {
      returnModule = true;
    }
    return this.deb(10, "*** START *** rootExports & noConflict() : exporting module '" + this.module.path + "' to root='" + rootName + "' & attaching `noConflict()`.") + (this.module.flags.noConflict ? ((function() {
      var _i, _len, _ref, _results;
      _ref = this.module.flags.rootExports;
      _results = [];
      for (i = _i = 0, _len = _ref.length; _i < _len; i = ++_i) {
        exp = _ref[i];
        _results.push("" + (i === 0 ? 'var ' : '    ') + "__old__" + exp + " = " + rootName + "." + exp);
      }
      return _results;
    }).call(this)).join(',\n') + ';\n' : '') + ((function() {
      var _i, _len, _ref, _results;
      _ref = this.module.flags.rootExports;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        exportedVar = _ref[_i];
        _results.push("" + rootName + "." + exportedVar + " = __umodule__");
      }
      return _results;
    }).call(this)).join(';\n') + ';\n' + (this.module.flags.noConflict ? "\n__umodule__.noConflict = " + this.__function(((function() {
      var _i, _len, _ref, _results;
      _ref = this.module.flags.rootExports;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        exp = _ref[_i];
        _results.push("" + rootName + "." + exp + " = __old__" + exp);
      }
      return _results;
    }).call(this)).join(';\n') + ';' + "\nreturn __umodule__;") + ';' : '') + '\n' + (returnModule ? "return __umodule__;" : '') + this.deb(10, "*** END *** rootExports & noConflict()");
  };

  Object.defineProperties(ModuleGeneratorTemplates.prototype, {
    factoryBodyAMD: {
      get: function() {
        return this.factoryBodyAll('AMD');
      }
    },
    factoryBodyNodejs: {
      get: function() {
        return this.factoryBodyAll('nodejs');
      }
    },
    _moduleExports_ModuleFactoryBody: {
      get: function() {
        return "module.exports = " + (this.__functionIIFE(this.module.factoryBody)) + ";";
      }
    }
  });

  ModuleGeneratorTemplates.prototype.factoryBodyAll = function(toKind) {
    if (toKind == null) {
      toKind = (function() {
        throw new Error("factoryBodyAll requires `toKind` in ['AMD', 'nodejs']");
      })();
    }
    return this.sp('useStrictModule', (!this.isCombined ? 'bundle.commonCode' : void 0), (!this.isCombined ? 'module.mergedCode' : void 0), 'module.beforeBody', ((toKind === 'nodejs') && (this.module.kind === 'AMD') ? ['_moduleExports_ModuleFactoryBody', "body of original '" + this.module.kind + "' module, assigned to module.exports"] : ['module.factoryBody', "body of original '" + this.module.kind + "' module"]), 'module.afterBody') + ((toKind === 'AMD') && (this.module.kind === 'nodejs') ? this.deb(20, 'returning nodejs `exports.module` as AMD factory result.') + "\nreturn module.exports;" : '');
  };

  ModuleGeneratorTemplates.prototype._genFullBody = function(fullBody) {
    var root;
    fullBody = this.sp('runtimeInfo', ['module.preDefineIIFEBody', 'statements/declarations before define(), enclosed in an IIFE (function(){})().']) + fullBody;
    return this.uRequireBanner + (this.module.isBare ? fullBody : this.module.isGlobalWindow ? (root = "(typeof exports === 'object' ? global : window)", this.__functionIIFE(fullBody, 'window', root, 'global', root)) : this.__functionIIFE(fullBody));
  };

  /*
  A Simple `define(['dep'], function(dep){...body...}})`,
  without any common stuff that are not needed for 'combined' template.
  
  `combined` template in AlmondOptimizationTemplate, merges/adds them only once.
  */


  ModuleGeneratorTemplates.prototype._AMD_plain_define = function() {
    return 'define(' + this.moduleNamePrint + this.defineArrayDepsPrint + this.__function(!this.isRootExports ? this.sp('factoryBodyAMD') : "var __umodule__ = " + this.__functionIIFE(this.sp('factoryBodyAMD'), this.parametersPrint, this.parametersPrint) + ";\n" + this._rootExportsNoConflict('window'), this.parametersPrint) + ')';
  };

  /*
    `combined` templates is actually defined in AlmondOptimizationTemplate.coffee,
    rendered through Bundle.coffee.
    When 'combined' is used, each `Module` in `Bundle` is converted as a `_AMD_plain_define`
  */


  ModuleGeneratorTemplates.prototype.combined = function() {
    return this._AMD_plain_define();
  };

  /* AMD template
      Runs only on WEB/AMD/RequireJs (and hopefully in node through uRequire'd *driven* RequireJS).
  */


  ModuleGeneratorTemplates.prototype.AMD = function() {
    return this._genFullBody(this._AMD_plain_define());
  };

  ModuleGeneratorTemplates.prototype.nodejs = function() {
    var dep, param, pi, prCnt;
    prCnt = 0;
    return this._genFullBody((_.any(this.module.nodeDeps, function(dep, depIdx) {
      var _ref;
      return (!dep.isSystem) && (((_ref = this.module) != null ? _ref.parameters : void 0) || [])[depIdx];
    }) ? "\nvar " : '') + ((function() {
      var _i, _len, _ref, _results;
      _ref = this.module.parameters;
      _results = [];
      for (pi = _i = 0, _len = _ref.length; _i < _len; pi = ++_i) {
        param = _ref[pi];
        if (!(dep = this.module.nodeDeps[pi]).isSystem) {
          _results.push((prCnt++ === 0 ? '' : '    ') + param + " = require(" + dep.name({
            quote: true
          }) + ")");
        }
      }
      return _results;
    }).call(this)).join(',\n') + ';' + this.sp('factoryBodyNodejs') + (this.isRootExports ? "var __umodule__ = module.exports;\n" + this._rootExportsNoConflict('global', false) : ''));
  };

  /*
    UMD template - runs AS-IS on both Web/AMD and nodejs (having 'npm install urequire').
    * Uses `NodeRequirer` to perform `require`s.
  */


  ModuleGeneratorTemplates.prototype.UMDplain = function() {
    return this.UMD(false);
  };

  ModuleGeneratorTemplates.prototype.UMD = function(isNodeRequirer) {
    var nDep, nr;
    if (isNodeRequirer == null) {
      isNodeRequirer = true;
    }
    nr = isNodeRequirer ? "nr." : "";
    return this._genFullBody(this.__functionIIFE((this.isRootExports ? "var rootExport = " + (this.__function(this._rootExportsNoConflict(), 'root, __umodule__')) + ";" : '') + ("if (typeof exports === 'object') {" + (isNodeRequirer ? "\n    var nr = new (require('urequire').NodeRequirer) ('" + this.module.path + "', module, __dirname, '" + this.module.webRootMap + "');" : '') + "\n    module.exports = " + (this.isRootExports ? 'rootExport(global, ' : '') + "factory(" + nr + "require" + this.injectExportsModuleParamsPrint + (((function() {
      var _i, _len, _ref, _results;
      _ref = this.module.nodeDeps;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        nDep = _ref[_i];
        if (nDep.isSystem) {
          _results.push(', ' + nDep.name());
        } else {
          _results.push(", " + nr + "require(" + (nDep.name({
            quote: true
          })) + ")");
        }
      }
      return _results;
    }).call(this)).join('')) + ")" + (this.isRootExports ? ')' : '') + ";\n} else if (typeof define === 'function' && define.amd) {\n    define(" + this.moduleNamePrint + this.defineArrayDepsPrint + (!this.isRootExports ? 'factory' : this.__function("return rootExport(window, factory(" + this.parametersPrint + "));", this.parametersPrint)) + ");\n  }"), 'factory', this.__function(this.sp('factoryBodyAMD'), this.parametersPrint)));
  };

  return ModuleGeneratorTemplates;

})(Template);

module.exports = ModuleGeneratorTemplates;
